### PR TITLE
Remove proto from commitments package

### LINKED ISSUE #717: Verifying a nil commitment silently passes
https://github.com/google/keytransparency/blob/master/core/client/kt/verify.go#L70:
```
// VerifyCommitment verifies that the commitment in `in` is correct for userID.
func (Verifier) VerifyCommitment(userID, appID string, in *tpb.GetEntryResponse) error {
	if in.Committed != nil {
		entry := new(tpb.Entry)
		if err := proto.Unmarshal(in.GetLeafProof().GetLeaf().GetLeafValue(), entry); err != nil {
			return err
		}
		if err := commitments.Verify(userID, appID, entry.Commitment, in.Committed); err != nil {
			return err
		}
	}
	return nil
}
```
If `in.committed == nil` the verification silently passes (no error is thrown). Is this intended? If so, why?
I would expect an error to be raised when Committed == nil

### PR DESCRIPTION
**Pull request by gdbelvin**
In order to make mobile bindings possible, we need to remove protobuf
references from verification interfaces.

Part of #715  #713
Closes #717

### CODE REVIEW

File: core/crypto/commitments/commitments.go [hunk 1]
```diff
@@ -53,26 +51,25 @@ var (
...
-	Write(ctx context.Context, commitment []byte, committed *tpb.Committed) error
+	Write(ctx context.Context, commitment, data, nonce []byte) error
 	// Read looks up a cryptograpic commitment and returns associated data.
-	Read(ctx context.Context, commitment []byte) (*tpb.Committed, error)
+	Read(ctx context.Context, commitment []byte) (data, nonce []byte, err error)
 }
 
 // Commit makes a cryptographic commitment under a specific userID to data.
-func Commit(userID, appID string, data []byte) ([]byte, *tpb.Committed, error) {
+func Commit(userID, appID string, data []byte) (commitment, nonce []byte, err error) {
```
Comment (AMarcedone) [inline core/crypto/commitments/commitments.go]: Despite being better than before (as it does not depend on protobuf), this function is still not gobind friendly, as it returns 3 outputs. To be "visible" from Java, functions need to return only one output (plus an optional second one which needs to be of type error).

**Comment by gdbelvin** [comment]
Would it be easier to return a struct, or should those be avoided as well?

**Comment by AMarcedone** [comment]
A struct would work (as long as its members are of gobind friendly types such as []byte).
I am not sure what is best from a design perspective, but my original idea was to leave the interface as it currently is and add some simple gobind friendly wrappers that internally call the original proto based interfaces. I think it is worth discussing this face to face tomorrow.

**Comment by AMarcedone** [comment]
Here is a concrete example of the interface I was thinking about (the B prefix is meant to indicate the goBind friendly version of a function)
https://github.com/AMarcedone/keytransparency/commit/6579ef03e5d478f158360ccf58c8d857f6b265a0

File: core/crypto/commitments/commitments.go (hunk 1 above)
Comment (gdbelvin) [inline core/crypto/commitments/commitments.go]: To fix this, I've created a `GenCommitmentKey` function.

File: core/keyserver/validate_test.go [hunk 2]
```diff
@@ -81,19 +81,23 @@ func TestValidateUpdateEntryRequest(t *testing.T) {
 	appID := "app"
 	vrfPriv, _ := p256.GenerateKey()
 	index, _ := vrfPriv.Evaluate(vrf.UniqueID(userID, appID))
-	commitment, committed, _ := commitments.Commit(userID, appID, profileData)
+	nonce, err := commitments.GenCommitmentKey()
```
Comment (cesarghali) [inline core/keyserver/validate_test.go]: Consider using a fixed nonce to make the test repeatable.

File: impl/sql/commitments/sql_test.go [hunk 3]
```diff
@@ -39,10 +39,11 @@ func TestWriteRead(t *testing.T) {
 
 	// Create test data.
 	pdata := []byte("key")
-	commitmentC, committedC, err := commitments.Commit("foo", "app", pdata)
+	nonce, err := commitments.GenCommitmentKey()
```
Comment (cesarghali) [inline impl/sql/commitments/sql_test.go]: Same here.

File: core/client/kt/verify.go [hunk 4]
```diff
@@ -88,9 +74,22 @@ func (Verifier) VerifyCommitment(userID, appID string, in *tpb.GetEntryResponse)
...
+
+	// If this is not a proof of absence, verify the connection between
+	// profileData and the commitment in the merkle tree leaf.
+	if in.GetCommitted() != nil {
+		commitment := entry.GetCommitment()
+		data := in.GetCommitted().GetData()
+		nonce := in.GetCommitted().GetKey()
+		if err := commitments.Verify(userID, appID, commitment, data, nonce); err != nil {
+			Vlog.Printf("✗ Commitment verification failed.")
+			return fmt.Errorf("VerifyCommitment(): %v", err)
```
Comment (gdbelvin) [inline core/client/kt/verify.go]: done

File: core/client/kt/verify_test.go [hunk 5]
```diff
@@ -1,73 +0,0 @@
-// Copyright 2016 Google Inc. All Rights Reserved.
```
Comment (gdbelvin) [inline core/client/kt/verify_test.go]: Good idea.

File: core/crypto/commitments/commitments.go [hunk 6]
```diff
@@ -50,36 +46,21 @@ var (
...
-type Committer interface {
-	// Write saves a cryptographic commitment and associated data.
-	Write(ctx context.Context, commitment []byte, committed *tpb.Committed) error
-	// Read looks up a cryptograpic commitment and returns associated data.
-	Read(ctx context.Context, commitment []byte) (*tpb.Committed, error)
-}
-
-// Commit makes a cryptographic commitment under a specific userID to data.
-func Commit(userID, appID string, data []byte) ([]byte, *tpb.Committed, error) {
+// GenCommitmentKey generates a commitment key for use in Commit.  This key
```
Comment (gdbelvin) [inline core/crypto/commitments/commitments.go]: done

File: core/keyserver/validate_test.go (hunk 2 above)
Comment (gdbelvin) [inline core/keyserver/validate_test.go]: done

File: impl/sql/commitments/sql_test.go (hunk 3 above)
Comment (gdbelvin) [inline impl/sql/commitments/sql_test.go]: done

**Comment by gdbelvin** [comment]
Added tests and cleaned up. PTAL.

File: core/client/kt/verify_test.go [hunk 7]
```diff
@@ -17,56 +17,106 @@ package kt
...
-			LeafProof: &trillian.MapLeafInclusion{
-				Leaf: &trillian.MapLeaf{
-					LeafValue: tc.entryData,
+		{
+			desc:    "empty",
+			userID:  "noalice",
+			appID:   "app",
+			trusted: &trillian.SignedLogRoot{},
+			in: &keytransparency_v1_types.GetEntryResponse{
+				VrfProof:  []byte{0x8, 0xca, 0x1, 0xaa, 0x1, 0x74, 0x70, 0xf4, 0xbf, 0x96, 0x69, 0xff, 0x73, 0x5d, 0xc7, 0x9d, 0x2b, 0x51, 0xb0, 0xa5, 0x42, 0xb8, 0x8c, 0x42, 0x34, 0xfc, 0xa1, 0xa2, 0xc0, 0x80, 0xdf, 0x76, 0x33, 0xfc, 0x7d, 0xee, 0x4a, 0xf3, 0x18, 0xea, 0x30, 0xc4, 0xa4, 0x6f, 0x31, 0xc9, 0x20, 0x73, 0x42, 0x84, 0xd9, 0x71, 0x39, 0x52, 0xb2, 0x8f, 0x58, 0x52, 0x4, 0x53, 0x87, 0x8, 0x3e, 0x81, 0x4, 0xb, 0x13, 0x89, 0xd7, 0xc6, 0x63, 0x22, 0x39, 0x18, 0x73, 0x72, 0xfa, 0x32, 0xf6, 0xeb, 0x3, 0x8, 0x5d, 0x7, 0x4e, 0x2, 0x3a, 0xc6, 0x7f, 0x89, 0xe8, 0x44, 0x27, 0xcb, 0x73, 0xdc, 0xf2, 0x2f, 0xcc, 0xcd, 0x90, 0x6e, 0x97, 0xcb, 0x22, 0xff, 0x6e, 0xdb, 0x74, 0x22, 0xbf, 0x28, 0x27, 0x9b, 0x9e, 0x26, 0x1a, 0xe4, 0xc6, 0x16, 0x59, 0x4f, 0x7d, 0xcc, 0xb9, 0x8e, 0x7d, 0x41, 0xf7},
```
Comment (cesarghali) [inline core/client/kt/verify_test.go]: Up to you: What do you think about expressing this as a hexa string and then convert to byte before using it? It'll be shorter. Same applies below.

Comment (gdbelvin) [inline core/client/kt/verify_test.go]: That's true, but makes the test vectors harder to regenerate with `log.Printf("%#v")`

Comment (cesarghali) [inline core/client/kt/verify_test.go]: :+1:

**Pull request merged**